src/EprPrnIntegration.Common/Service/ServiceBusProvider.cs

```diff
@@ -1,45 +1,223 @@
+        ILogger<ServiceBusProvider> logger,
+        ServiceBusClient serviceBusClient,
+        IOptions<ServiceBusConfiguration> config) : IServiceBusProvider
     {
-        private readonly ILogger<ServiceBusProvider> _logger;
-        private readonly ServiceBusClient _serviceBusClient;
-        private readonly IOptions<ServiceBusConfiguration> _serviceBusConfig;
-        public ServiceBusProvider(ILogger<ServiceBusProvider> logger, ServiceBusClient serviceBusClient, IOptions<ServiceBusConfiguration> config)
-        {
-            _logger = logger;
-            _serviceBusClient = serviceBusClient;
-            _serviceBusConfig = config;
-        }
         public async Task SendFetchedNpwdPrnsToQueue(List<NpwdPrn> prns)
         {
+            ServiceBusMessageBatch? messageBatch = null;
             try
             {
-                await using var sender = _serviceBusClient.CreateSender(_serviceBusConfig.Value.FetchPrnQueueName);
-                using ServiceBusMessageBatch messageBatch = await sender.CreateMessageBatchAsync();
+                await using var sender = serviceBusClient.CreateSender(config.Value.FetchPrnQueueName);
+                messageBatch = await sender.CreateMessageBatchAsync();
                 foreach (var prn in prns)
                 {
                     var jsonPrn = JsonSerializer.Serialize(prn);
-                    if (!messageBatch.TryAddMessage(new ServiceBusMessage(jsonPrn)))
+                    var message = new ServiceBusMessage(jsonPrn);
+                    if (!messageBatch.TryAddMessage(message))
+                    {
+                        logger.LogInformation("SendFetchedNpwdPrnsToQueue - Batch overflow sending this batch with message count {count}", messageBatch.Count);
+                        await sender.SendMessagesAsync(messageBatch);
+
+                        logger.LogInformation("SendFetchedNpwdPrnsToQueue - Disposing current batch and creating new batch");
+                        messageBatch.Dispose();
+                        messageBatch = await sender.CreateMessageBatchAsync();
+
+                        logger.LogInformation("SendFetchedNpwdPrnsToQueue - Adding message in new batch");
+                        if (!messageBatch.TryAddMessage(message))
+                        {
+                            throw new InvalidOperationException("SendFetchedNpwdPrnsToQueue - Could not add message into new batch");
+                        }
+                    }
+                }
+                if (messageBatch.Count > 0)
+                {
+                    logger.LogInformation("SendFetchedNpwdPrnsToQueue - Sending final batch with message count {count}", messageBatch.Count);
+                    await sender.SendMessagesAsync(messageBatch);
+                }
+                logger.LogInformation("SendFetchedNpwdPrnsToQueue - total {MessageBatchCount} messages has been published to the queue: {queue}", messageBatch.Count, config.Value.FetchPrnQueueName);
```

Comment:
Being picky but you called this {count} above and now the same variable is {MessageBatchCount}

Reply:
Kudos to your eagle eyes :). I have changed it, even though it is ok to have different parameter names :D